### DISCUSSION TITLE
Device Node (alpha)

### DISCUSSION THREAD

**Discussion by zachowj** [zachowj/node-red-contrib-home-assistant-websocket#367] [General]
Placeholder

**Comment by nmsousa76** [comment]
None of my ZHA lights are showing up as devices nodes in NodeRed.   All other lights show up fine.  (zwave, tuya, hue).   Is this something specific to ZHA

**Reply by zachowj** [reply]
I only have two ZHA sensors and no lights so I can't really test it. It uses the same API as the automations in HA so if they show there they should show up in NR.

You could try this test flow to see if HA API is actually returning them or not.

```
[{"id":"3e866583.beefda","type":"inject","z":"ffbd7f06.4a014","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":172,"y":992,"wires":[["368d5f81.4acc"]]},{"id":"c84bede7.8a01","type":"debug","z":"ffbd7f06.4a014","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":678,"y":992,"wires":[]},{"id":"368d5f81.4acc","type":"ha-api","z":"ffbd7f06.4a014","name":"","server":"","version":1,"debugenabled":false,"protocol":"websocket","method":"get","path":"","data":"{\"type\": \"config/device_registry/list\"}","dataType":"jsonata","responseType":"json","outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"results"}],"x":336,"y":992,"wires":[["3f8c0f6b.e8877"]]},{"id":"3f8c0f6b.e8877","type":"function","z":"ffbd7f06.4a014","name":"","func":"msg.payload = msg.payload.filter(entity => entity.identifiers && entity.identifiers.length && entity.identifiers[0][0] === \"zha\");\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":508,"y":992,"wires":[["c84bede7.8a01"]]}]
```

**Reply by nmsousa76** [reply]
Devices show up on on flow, but not on node object.  See screenshot

![image](https://user-images.githubusercontent.com/65356035/119728521-9496e200-be41-11eb-9b50-757c8817d74d.png)

![image](https://user-images.githubusercontent.com/65356035/119728609-ad9f9300-be41-11eb-9d7a-7303f5e930b3.png)

**Reply by nmsousa76** [reply]
I think I figured out what is wrong, but would need someone to fix it.   It looks like the node is using "name" rather than "name_by_user". As you can see all my ledvance bulbs are showing up as the same name.  As you can see this would be problematic for many devices of the same brand.   This is not specific to lights either as I noticed the same issue on my Orbit faucets.  


![image](https://user-images.githubusercontent.com/65356035/119730103-58fd1780-be43-11eb-82a8-9fa2490fef87.png)

**Reply by zachowj** [reply]
Didn't know `name_by_user` was a thing. I'll have to look into that.

**Reply by zachowj** [reply]
It will use `name_by_user` if it exists in the next release

**Reply by nmsousa76** [reply]
Thank you

**Comment by joryvd** [comment]
I have some Ikea tradfri wireless dimmers connected through Zigbee. I am using them with the ZHA integration in Home Assistant. When I listen to ZHA events and output this in debug I can see the messages coming in.
When I add them as a device --> trigger --> turn on pressed and connect this to the debug window nothing comes up. Not for any of the button presses that come out of it. I tried an automation via the Home Assistant GUI using the same triggers and that works. So something is not getting passed through to the device in Node-red. If any further info is needed I'm glad to supply.

**Reply by zachowj** [reply]
Does the property `event` on the device node look the same as the automation in HA?

![image](https://user-images.githubusercontent.com/37859597/129972879-f429b288-6efb-4573-8456-db6a8fa9baf9.png)

![image](https://user-images.githubusercontent.com/37859597/129972899-1ff274bc-2898-4a25-9af3-f8eb87b5d415.png)

**Reply by nilsi** [reply]
Got the same issue with an Ikea Tradfri On/Off button. Trying to trigger on "Turn off pressed" event but it doesn't come through to node red. 

I can see the event triggered in HA:

<img width="1038" alt="image" src="https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/793881/13a42c8d-640e-4e10-9685-2ec37fecfdf6">

But nothing in node-red when I connect a debug node:
<img width="1351" alt="image" src="https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/793881/84ad80b1-0fdf-4ba0-b468-f7f0e42de879">

**Reply by nilsi** [reply]
If I connect a debug node to events all nodes I can see this event in node-red so it gets an event at least.

<img width="404" alt="image" src="https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/793881/2eb64843-43e1-46e2-97e1-4832d0f4ac16">

**Comment by SpeedyDog14** [comment]
Hi zachowj,

Do you have any plans to 

1) Add option to output on any trigger    
2) Add option to have multiple outputs with selectable trigger options against each output

Thanks in advance

**Reply by zachowj** [reply]
> 1. Add option to output on any trigger

Probably not because all the logic is handled by HA for the device node.

> 2. Add option to have multiple outputs with selectable trigger options against each output

Initially, I had planned to be able to register multiple device actions under one node but wanted to test out a single one to start. I think I had planned to have them all output out a single node not multiple. Do you think multiple outputs would be better? I feel if you wanted multiple outputs you would using multiple nodes.

**Reply by SpeedyDog14** [reply]
Hi zachowj and thanks for your reply,

My example use case is where I have a Zigbee Aqara Opple 6 button battery powered stateless remote switch.

![image](https://user-images.githubusercontent.com/64558508/134390811-ec5453d6-cb84-409b-be3e-a163b9b39f84.png)

Currently in Node-Red I can capture the Zigbee event for this particular type of Switch in one ZHA Event Node that then output to a single Switch Node with 6 outputs (one output for each of the 6 buttons) I can then add a suitable Action/service Node to the output of each of the node switch outputs.

The only way I can see to do this with your new Device Node is to have a separate Device Node for each trigger (button press). This becomes untidy especially if I decide to expand and fully utilise double/triple clicks/hold actions of each button. I would potentially need 24 Device Nodes to cater for all the triggers instead of just one.

If you can't cater to trigger on any device event then it is what it is and is really not an issue for me. I was just wondered what your road map was so I can make the right decisions moving forward with my flows.

Cheers

**Reply by zachowj** [reply]
At this point, I would recommend staying with what you have. Any changes I plan to make to the device are not going to simplify how you are currently doing things.

**Comment by dirk2947** [comment]
Hey there, just wanted to say thank you. Came across it by chance and it was super intuitive with my aliexpress 4 button zigbee switches.

**Comment by jlsjonas** [comment]
Having some strange behavior using the new tradfri buttons with it...
![image](https://user-images.githubusercontent.com/932193/143055168-50b30953-01ba-41e1-a8e5-58d1e4790c02.png)
a single tap seems to consistently equal to 3 events in NODE-RED :thinking: Anything I could check to help? (before this it was hooked to a regular HA automation, where I didn't experience this behavior
switch is connected via ZHA

**Reply by zachowj** [reply]
Is the payload data the exact same in all three events?

**Reply by jlsjonas** [reply]
:thinking: almost... everything is exactly the same, apart from the `payload.event.context.id` (ids remain the same every time I click)
I've just put a 50ms debounce after it for my purposes (although that does cause some issues with double-tap handling), but it still feels like something to investigate before moving the device node to stable :)

 -- let me know if I can help out here

**Reply by redhawk07** [reply]
I have the same issue, but the event is fired 5 times.
The data of the message object only differs in these parts:

- msg.event.time_fired
- msg.event.context.id
- msg._msgid

Everything else of the msg object is 100 % same.
Using version 0.43.0 of node-red-contrib-home-assistant-websocket.

**Comment by han9over** [comment]
I'm trying to use Action type to send a notification to my phone and I can do that by hardcoding Title and Message. However, I want to use single node and dynamically change the title and message used. How can I achieve that as I can't use msg.payload

**Reply by zachowj** [reply]
Currently, this is not possible with the device node. You should be able to call service node in its place.

**Reply by han9over** [reply]
Thanks. Will try that out.

**Comment by Apflkuacha** [comment]
I really find it great this Device Node, but as @SpeedyDog14 already mentioned an option to output all triggers. In my case I do use a function to react to events from a remote button and all the triggers should be forwarded to my function, the information what was pressed is in the payload.
If I do use the events:state Node I do get a single click and double click event when I only do a single click, so I prefer this Device Node.

**Comment by motionist** [comment]
When using the device for a Styrbar switch (Connected through deConz), the text does not match the actual buttons - instead the assignment is according to the codes sent by deConz (1002, 1001, 1003, 2002, 2001, 2003, 3002, 3001, 3003, 4002, 4001, 4003) - basically having the switch with Increase brightness/turn on upwards the order is anti-clock wise

**Reply by zachowj** [reply]
NR uses data for the device node directly from Home Assistant. Test it as a Home Assistant automation device trigger and see if you get the same results. If so, then you would need to report to the repo of the deconz integration.

**Comment by bravehand** [comment]
First off, thanks for this!  I'm using it with some homeseer switches to trigger on central scene activation but getting some odd behavior in the "Value" field.  When I create/save the node it's set to the 4th option in the dropdown (KeyPressed2x):
![image](https://user-images.githubusercontent.com/21265469/163829949-c677ff2f-5f82-48fe-9d4a-7ee3f39ec59e.png)

but when i open the node again it's set to the first value in the dropdown (KeyPressed):
![image](https://user-images.githubusercontent.com/21265469/163830223-0f7c0fab-6253-46a0-b338-29ae997c90ff.png)

If I don't reopen the node after saving, things seem to work ok for a while but eventually start triggering on the first option (KeyPressed).  I haven't tested to confirm, but I suspect it's on a restart of HA and/or node-red.

And this has to potentially to really simplify central scene triggers, just look at how I'm doing it without this node:
![image](https://user-images.githubusercontent.com/21265469/163829313-bad5e262-6265-49a3-8b5c-ab8dce8441c5.png)

**Comment by mltiede** [comment]
I'm fairly new to Synology, reasonably new to Home Assistant (thanks to Insteon), and totally new with Node-Red.  I got Node-Red installed in a docker container and am able to run it.  Then I added HACS to Home Assistant and added the HA part for Node-Red.  Now when I try to put a node on the Node-Red panel, and go to Edit Device Node and Edit Server Node dialog and fill in the IP address and port to Home Assistant and press Update, I get an error that says "Error retrieving translations".

I presume this is something to do with permissions in the node-red folder on the NAS.  I see that "root" is the owner.  Should I make it something else somehow?  Or is something else going on?

... time passes ...

I changed owner from root to my adminstrator user.  Still doesn't work.  (and I don't see how to change it back to "root" in case that matters)

... time passes ...

FWIW, when I entered the IP and port in the browser, I get a prompt for user and password and then for 2 factor authentication.  Could that be causing a problem?  Node-Red isn't prompting and doesn't provide a place to put that information.

... time passes ...

How can I tell what user Home Assistant and Node-Red are using in their docker containers?  I see that I set up a special Home Assistant User, but I must have determined that it wasn't being used because its permissions are DENY for the config and subfolders.  I tried giving it full permissions, but it didn't make any difference.

**Comment by TrickShow** [comment]
I just wanted to say this is awesome and has been working so well for me. Thank you for all that you do!

**Comment by johan-beemsterboer** [comment]
I'm struggling to get the node working, anyone here who can help me? 

I have a remote here ('Afstandsbediening') with 8 buttons. The node says it's running:

![image](https://user-images.githubusercontent.com/109644704/180142692-e6f0f673-ccb8-4d11-b19c-cb793327f989.png)


But when I push the button, nothing happens. What am I doing wrong?

![image](https://user-images.githubusercontent.com/109644704/180142854-6fd98362-1fe8-4fa9-8dd6-1ca8cad10342.png)

**Reply by zachowj** [reply]
Do you have the custom component installed in HA? It should say "registered" under the node after you deploy if everything is set up correctly.

**Reply by johan-beemsterboer** [reply]
Been a while, but managed to fix it with installing the custom component. That was the one, thanks!

**Comment by tamueller** [comment]
I'm new to node red, but I can't get this to work. I installed node-red in homeassistant and use zigbee2mqtt. When using a state_changed node I get the events from z2m, but not for the device node. What am I doing wrong?
![image](https://user-images.githubusercontent.com/1902960/193593403-ea25095d-f6ed-4664-a418-19ee936560d6.png)

As you can see, debug 7 gets fired, debug 6 not.

**Reply by zachowj** [reply]
The device node requires the NR custom component. https://github.com/zachowj/hass-node-red

When that is installed in HA the status text of the device node will show 'registered' and this 
![image](https://user-images.githubusercontent.com/37859597/193594369-75a0fc93-b3ce-48e7-bbda-96d5cdff6209.png)
 will go away from the device editor.

**Reply by tamueller** [reply]
Thanks, that was the problem.

**Comment by tamueller** [comment]
@zachowj: The device node is working mostly fine now, but in my case all of the device triggers show up in HA as a switch, which clutters my auto-generated dashboard. Is it possible to add a switch, whether to expose it to HA, and set the default to false. I can see some cases where this might be useful, but I think in most cases, you don't need the trigger to be exposed to HA.

**Comment by LordMike** [comment]
This device node is awesome. The help provided by autocompletions and dropdowns is great and greatly simplifies the troubleshooting part of my setup..

In order to better use f.ex. function nodes (because I'd like to do some custom scripting), I find myself needing to be able to react to multiple devices, multiple triggers etc.. in order to handle the events in my own code. So my suggestion is that the `Value`, `Trigger` or `Device` dropdowns support multiple values, that'd be awesome.

I'm intending to use it for multiple devices of the same make/model, so my Trigger/Value dropdowns will be identical, but I have no idea how to make great UX with different sets of values... Maybe only show values/triggers that overlap?

So I have two use cases really:
* I'd like to trigger a flow, when a specific button is pressed on one of many switches - ie., I want to select multiple devices, but keep the same Trigger/Value combination
* I'd like to trigger a flow, when a specific button is pressed either once or twice on one of my switches - ie., I want to select multiple Value's (`KeyPress`, `KeyPress2`, `...`), for one trigger.

Example, these three trigger the one the _same_ device and trigger, but for different values. And I'd like to duplicate this for about 4 or 5 switches, so that currently makes for ~15 nodes total.. :)
![image](https://user-images.githubusercontent.com/1027111/203031568-219837ce-97ac-41ee-824a-48def011d824.png)

**Comment by ivanlofgren** [comment]
### Device Node crash Node Red when selecting unavailable device

**Recreation procedure**
Add device node to node red
Edit the node and select an unavailable device (for me eg shelly plug that's been offline for a long time)
This crashes node red integration, which then restart.
During restart, opening device nodes gives translation error message below. During restart, node red publishing also isn't possible, resulting in a 502 bad gateway error.

![image](https://user-images.githubusercontent.com/73953790/204006428-0e881540-1654-498a-9934-368264077393.png)

**Comment by LordMike** [comment]
If I make a device trigger on the `Node status changed`, it seems I _must_ choose a _from_ and _to_. So this means if I just want to know if a device becomes dead/unavailable, I have to make 2-3 triggers, for each `From-To` pair?

Other Node Red nodes use blank values to indicate that the value is irrelevant. 

![image](https://user-images.githubusercontent.com/1027111/204092058-b352757e-cd42-4a74-8293-4e60903db8d6.png)

**Reply by zachowj** [reply]
Yes, this should be added.

**Comment by bednim** [comment]
Is DEVICE node the to use when it comes to binary_sensor (like contact sensors etc etc)? I have noticed these are now gone from event_state. thanks!

**Comment by DennisGaida** [comment]
Using the **device** node for the first time. Do I need to use multiple nodes for multiple triggers?

I have one device that has triggers such as **on**/**off**/**up**/**down** and **stop**. I'd love to use a switch node to do different things depending on trigger/button press. The device node seems to only allow one trigger?

**Reply by zachowj** [reply]
Yes, you need one node for each trigger.

**Reply by DennisGaida** [reply]
Thanks for your answer, thought as much. I don't like device triggers then ;-) I suppose because you have to register for each handler with HA to receive events? Wish we could register for **(*)** then...

**Comment by catduckgnaf** [comment]
Under Device Node, "Brightness Increase" and "Brightness Decrease" do not work. I was looking for an easy way out of adjusting brightness. Is this known issue?

**Comment by yameiko** [comment]
since installation of hacs update node companion: node with aqara switch devices doesnt work anymore. ha automation with this devices still work. does anyone have the same problem

**Reply by zachowj** [reply]
If you're not running v0.57.0 beta of the HA nodes revert back to version 2.2.0

**Reply by catduckgnaf** [reply]
Is this a planned breaking change? Or will it be fixed? And what should we do instead? (Other then not update)

**Reply by zachowj** [reply]
https://github.com/zachowj/hass-node-red/releases/tag/v3.0.0

**Reply by catduckgnaf** [reply]
I guess I meant, is there a solution? Other then staying on an old version. If that makes sense (or not using node-red for these devices?)

**Reply by tangrasmus** [reply]
Had the same problem with aqara buttom uninstalled 3.0.0 went back to 2.2.0 and it works again.

**Comment by JonnySpruce** [comment]
I have a presence sensor set up to trigger when the attribute changes to "Approch". Using this node it's multiple seconds slower than if using a HA automation. Is that expected and is there anything I can do about it?
